cli: add namespace consistency check to doctor tool

spaskob opened the pull request:
So far the doctor tool would run `Validate` on all table descriptors
and report any validation errors. In this PR we ingest the namespace
table as well (from `sysstem.namespace2.txt or from a live cluster).
The namespace entries are then validated against the descriptors:
- each descriptor that is not dropped should be pointed to by at least
one namespace row.
- the name of the table should be in found in a namespace row that points
to the table descriptor.
- no NULL values should be found in the namespace table.
- every namespace row should point to an existing descriptor.

Touches #51153.

Release note: none.

Release justification: this is a debug tool that examines a debug zip or
a cluster for problems. There are no changes to the DB itself.

Commit:
cli: add namespace consistency check to doctor tool

So far the doctor tool would run `Validate` on all table descriptors
and report any validation errors. In this PR we ingest the namespace
table as well (from `sysstem.namespace2.txt or from a live cluster).
The namespace entries are then validated against the descriptors:
- each descriptor that is not dropped should be pointed to by at least
one namespace row.
- the name of the table should be in found in a namespace row that points
to the table descriptor.
- no NULL values should be found in the namespace table.
- every namespace row should point to an existing descriptor.

Touches #51153.

Release note: none.

Release justification: this is a debug tool that examines a debug zip or
a cluster for problems. There are no changes to the DB itself.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/53465)
<!-- Reviewable:end -->

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/53465)
<!-- Reviewable:end -->

rohany commented on pkg/cli/doctor.go:
```diff
@@ -166,7 +163,61 @@ func runZipDirDoctor(cmd *cobra.Command, args []string) (retErr error) {
+		}
+		id, err := strconv.Atoi(fields[3])
+		if err != nil {
+			if fields[3] == "NULL" {
```
id being null is already a problem, we should error out here

spaskob replied:
this code just collects the data and the function `Examine` reports any problems. I don't stop here so that we can try to report as many problems as possible.

rohany commented on pkg/cli/doctor.go:
```diff
@@ -166,7 +163,61 @@ func runZipDirDoctor(cmd *cobra.Command, args []string) (retErr error) {
+}
+
+// tableMap applies `fn` to all rows in `in`.
+func tableMap(in io.Reader, fn func(string) error) error {
```
why can't you use go's csv reader or something?

spaskob replied:
thank you - of course!

rohany commented on pkg/sql/doctor/doctor.go:
```diff
@@ -46,32 +56,121 @@ func NewDescGetter(rows []DescriptorTableRow) (catalog.MapDescGetter, error) {
-	fmt.Fprintf(stdout, "Examining %d descriptors...\n", len(descTable))
-	descGetter, err := NewDescGetter(descTable)
+func Examine(
+	descTable []DescriptorTableRow,
```
[nit]: might be clean to alias these types as well, like `DescriptorTable`

rohany commented on pkg/sql/doctor/doctor.go:
```diff
@@ -46,32 +56,121 @@ func NewDescGetter(rows []DescriptorTableRow) (catalog.MapDescGetter, error) {
-		if !isTable {
-			// So far we only examine table descriptors. We may add checks for other
-			// descriptors in later versions.
+		if isTable {
```
[nit]: I'd prefer doing a switch on the type here. It would be worthwhile to add validation for types and schemas as well here, since we are just adding them this release

spaskob replied:
Good point - could you describe what needs to be validated for type and schema descriptors?

rohany replied:
Types have the same validation as tables (parentID and parent schema ID should be valid). Schema descriptors should have parentSchemaID be 0, and a valid parent ID

rohany commented on pkg/sql/doctor/doctor.go:
```diff
@@ -46,32 +56,121 @@ func NewDescGetter(rows []DescriptorTableRow) (catalog.MapDescGetter, error) {
+			}
+		}
+
+		// Process namespace entries point1ing to this descriptor.
```
/s/point1ing/pointing/

rohany commented on pkg/sql/doctor/doctor.go:
```diff
@@ -46,32 +56,121 @@ func NewDescGetter(rows []DescriptorTableRow) (catalog.MapDescGetter, error) {
+	// Now go over all namespace entries that don't point to descriptors in the
+	// descriptor table.
+	for id, ni := range nMap {
+		if id == 29 {
```
keys.PublicSchemaID

rohany commented on pkg/sql/doctor/doctor.go:
```diff
@@ -46,32 +56,121 @@ func NewDescGetter(rows []DescriptorTableRow) (catalog.MapDescGetter, error) {
-			fmt.Fprintf(stdout, "Table %3d: validated\n", table.GetID())
+			continue
 		}
+		fmt.Fprintf(
```
Temporary schemas will have namespace entries but not descriptors (for now). You can handle this by checking `strings.HasPrefix(name, "pg_temp_")`

spaskob commented:
PTAL

rohany commented on pkg/cli/doctor.go:
```diff
@@ -86,22 +89,12 @@ func runClusterDoctor(cmd *cobra.Command, args []string) (retErr error) {
 	defer sqlConn.Close()
 
-	rows, err := sqlConn.Query(`
+	stmt := `
```
do you need a version check to use `crdb_internal_mvcc_timestamp` here?

spaskob replied:
yes, how do I do that exactly (any similar examples?)

rohany replied:
One way that I've done this is to first issue a query like `select crdb_internal_mvcc_timestamp from ... limit 1` and see if an `UndefinedColumn` error is returned.

spaskob replied:
Thanks!

spaskob replied:
I realize I have similar problem when using `system.namespace2`. Do I need to worry about it if I only want to support clusters >= 20.1? Going forward, which table are going to keep `system.namespace` or `system.namespace2`?

rohany replied:
after a cluster hits 20.1 the main namespace table to consider is `system.namespace`, (it will be the one with the schema column)

rohany replied:
You will have to check that the column `parentSchemaID` exists probably

rohany commented on pkg/sql/doctor/doctor.go:
```diff
@@ -46,32 +62,140 @@ func NewDescGetter(rows []DescriptorTableRow) (catalog.MapDescGetter, error) {
+		desc, ok := descGetter[descpb.ID(row.ID)]
+		if !ok {
+			// This should never happen as ids are parsed and inserted from descTable.
+			log.Fatalf(context.Background(), "Descriptor id %d not found", row.ID)
```
you should just pass in a context into examine, there are alot of context users here now

rohany commented on pkg/sql/doctor/doctor.go:
```diff
@@ -46,32 +62,140 @@ func NewDescGetter(rows []DescriptorTableRow) (catalog.MapDescGetter, error) {
+			}
+			// Table has been already validated.
+			parentExists = true
+			parentSchemaExists = true
```
missing type descriptors here? They have a validate method as well

spaskob replied:
I could not find it in `descriptor.go`. Where is it defined?

rohany replied:
look at `type_desc.go`

spaskob replied:
done

spaskob reviewed (commented):
PTAL

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53465#-:-MG9mEyh91P-as44-yEF:bb71pro)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lucy-zhang, @rohany, and @spaskob)



<!-- Sent from Reviewable.io -->

spaskob commented:
PTAL

Commit:
cli: support versions prior to 20.1

The data extraction code from a live cluster uses
`crdb_internal_mvcc_timestamp` which was introduced in 20.2. For prior
versions just use the current timestamp.

Release note: none.

Release justification: none

spaskob commented:
PTAL

spaskob reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53465)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lucy-zhang and @rohany)

___
*[pkg/cli/doctor.go, line 189 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53465#-MFdqW3NGgvHaabJOcp3-r1-189:-MGENkN51lihPiOy_Vo0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac7e3153b6ac4f82dbc19b4dfb759acc95aa72e8/pkg/cli/doctor.go#L189)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

this code just collects the data and the function `Examine` reports any problems. I don't stop here so that we can try to report as many problems as possible.
</blockquote></details>

Done.
___
*[pkg/cli/doctor.go, line 211 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53465#-MFdqW3NGgvHaabJOcp3-r1-211:-MGENlFtAbT9XKttuVNR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac7e3153b6ac4f82dbc19b4dfb759acc95aa72e8/pkg/cli/doctor.go#L211)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

thank you - of course!
</blockquote></details>

Done.
___
*[pkg/sql/doctor/doctor.go, line 74 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53465#-MFdqW3NGgvHaabJOcp4-r1-74:-MGENmNdA4qwHNYeuxZp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac7e3153b6ac4f82dbc19b4dfb759acc95aa72e8/pkg/sql/doctor/doctor.go#L74)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

[nit]: might be clean to alias these types as well, like `DescriptorTable`
</blockquote></details>

Done.
___
*[pkg/sql/doctor/doctor.go, line 103 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53465#-MFdqW3NGgvHaabJOcp4-r1-103:-MGENoPWD0wd4oY2WThg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac7e3153b6ac4f82dbc19b4dfb759acc95aa72e8/pkg/sql/doctor/doctor.go#L103)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

Types have the same validation as tables (parentID and parent schema ID should be valid). Schema descriptors should have parentSchemaID be 0, and a valid parent ID
</blockquote></details>

Done.
___
*[pkg/sql/doctor/doctor.go, line 117 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53465#-MFdqW3NGgvHaabJOcp4-r1-117:-MGENpit4Ki-gnTYnF80:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac7e3153b6ac4f82dbc19b4dfb759acc95aa72e8/pkg/sql/doctor/doctor.go#L117)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

/s/point1ing/pointing/
</blockquote></details>

Done.
___
*[pkg/sql/doctor/doctor.go, line 152 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53465#-MFdqW3NGgvHaabJOcp4-r1-152:-MGENqDW8kbVAVzXQtZV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac7e3153b6ac4f82dbc19b4dfb759acc95aa72e8/pkg/sql/doctor/doctor.go#L152)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

keys.PublicSchemaID
</blockquote></details>

Done.
___
*[pkg/sql/doctor/doctor.go, line 160 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53465#-MFdqW3NGgvHaabJOcp4-r1-160:-MGENqiw56IpxhlfPgHK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac7e3153b6ac4f82dbc19b4dfb759acc95aa72e8/pkg/sql/doctor/doctor.go#L160)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

Temporary schemas will have namespace entries but not descriptors (for now). You can handle this by checking `strings.HasPrefix(name, "pg_temp_")`
</blockquote></details>

Done.
___
*[pkg/sql/doctor/doctor.go, line 96 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/53465#-MFdqW3NGgvHaabJOcp4-r2-96:-MGENnCg6gDixUoUUph1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/98c0b092dcff0af66116bc6876e994ad8d03cade/pkg/sql/doctor/doctor.go#L96)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

you should just pass in a context into examine, there are alot of context users here now
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rohany reviewed (approved):
lgtm aside from the namespace related comment above

thoszhang reviewed (approved):
LGTM so far. Can I suggest that we also check that the draining names list on the descriptor matches the contents of the namespace table? Basically, every namespace entry that is does not correspond to the current name of the table should be in the list of draining names. If you want to do it in a follow-up PR that's also fine.

> Reviewed 1 of 6 files at r2, 4 of 5 files at r3, 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53465#-:-MGHB1iu3X9n4ET2rlmo:b-bx3ddf)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rohany and @spaskob)



<!-- Sent from Reviewable.io -->

spaskob reviewed (commented):
Thanks Lucy, I will send a follow up PR soon for this check.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53465#-:-MGITVoo5Y2ancso4uTL:bp4qrbj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rohany)

___
*[pkg/cli/doctor.go, line 92 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/53465#-MFdqW3NGgvHaabJOcp3-r2-92:-MGITPHf41pRm0TnoqPH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/98c0b092dcff0af66116bc6876e994ad8d03cade/pkg/cli/doctor.go#L92)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

You will have to check that the column `parentSchemaID` exists probably
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

spaskob commented:
bors r+
